apps/openmw/mwmechanics/spellcasting.cpp

```diff
@@ -763,6 +763,35 @@ namespace MWMechanics
 
         int school = 0;
 
+        if(mCaster.getClass().isActor()
+                && MWBase::Environment::get().getWorld()->getStore().get<ESM::MagicEffect>()
+                .find(ESM::MagicEffect::WaterWalking))
+        {
+            const MWWorld::CellStore *cell = mCaster.getCell();
+            
+            if(cell->getCell()->hasWater())
+            {
+                const ESM::Position casterPosition = mCaster.getRefData().getPosition();
+                const float casterHeight = mCaster.getCellRef().getScale();
+
+                if(MWBase::Environment::get().getWorld()->castRay(
+                        casterPosition.pos[0], casterPosition.pos[1], casterPosition.pos[2],
+                        casterPosition.pos[0], casterPosition.pos[1], cell->getWaterLevel() + casterHeight ))
+                {
+                    // not enough space to place caster
+                    MWBase::Environment::get().getWindowManager()->messageBox ("#{sMagicInvalidEffect}");
+                    return 0;
+                }
+                else if(MWBase::Environment::get().getWorld()->isUnderwater(cell, casterPosition.asVec3()))
+                {
+                    // move the caster to surface
+                    ESM::Position newCasterPosition = casterPosition;
+                    newCasterPosition.pos[2] = cell->getWaterLevel();
+                    mCaster.getRefData().setPosition(newCasterPosition);
```

Comment:
Is this even working? To correctly move an actor, you need to use World::moveObject.